programs/marinade-finance/src/state/mod.rs

```diff
@@ -255,13 +261,46 @@ impl State {
         Ok(())
     }
 
-    /*
-    pub fn register_stake_order(&mut self, lamports_amount: u64) {
-        self.epoch_stake_orders += lamports_amount;
+    // ---------------
+    // EMERGENCY PAUSE
+    // ---------------
+
+    // is paused if `MAX_PAUSE_EPOCHS` haven't elapsed starting from the last `pause()`
+    // is paused if we haven't reached `resume_at_epoch`
+    // as soon as we enter `resume_at_epoch` epoch the contract will self-resume
+    pub fn is_paused(&self) -> Result<bool> {
+        Ok(Clock::get()?.epoch < self.resume_at_epoch)
+    }
+
+    // set resume_at_epoch to current epoch
+    pub fn pause(&mut self) -> Result<()> {
+        require!(!self.is_paused()?, MarinadeError::AlreadyPaused);
+        let epoch = Clock::get()?.epoch;
+        // if we just ended the pause, and for the next epoch, can't re-pause
+        require_gt!(
+            epoch,
+            self.resume_at_epoch + 1,
+            MarinadeError::TooSoonToRePause
+        );
+        self.resume_at_epoch = epoch + State::PAUSE_MAX_EPOCHS;
+        Ok(())
     }
 
-    pub fn register_unstake_order(&mut self, unstaker_index: u32, lamports_amount: u64) {
-        self.epoch_unstake_orders += lamports_amount;
-        self.unstakers[unstaker_index as usize].amount += lamports_amount;
-    }*/
+    pub fn resume(&mut self) -> Result<()> {
+        require!(self.is_paused()?, MarinadeError::NotPaused);
+        // set resume_at_epoch to current epoch
+        // this unpauses immediately
+        // and prevents re-pausing for the current epoch and the next
+        self.resume_at_epoch = Clock::get()?.epoch;
+        Ok(())
+    }
+
+    // returns MarinadeError::ProgramIsPaused if paused
+    pub fn check_paused(&self) -> Result<()> {
```

Comment:
The name must work backwards like "check_working". Better to inline require!(!self.is_paused(), MarinadeError::ProgramIsPaused) instead of a separate function to have less jumps for code reader

Reply:
can make the fn explicitly inline, but the two ! ! are annoying. let me think about a #[pausable] decorator

Reply:
name changed